components/TextFields/src/MDCTextInputController.m

```diff
@@ -0,0 +1,1201 @@
+                                   context:nil];
+  [_textInput.trailingUnderlineLabel addObserver:self
+                                      forKeyPath:MDCTextInputControllerKVOKeyFont
+                                         options:0
+                                         context:nil];
+  _isRegisteredForKVO = YES;
+}
+
+- (void)unsubscribeFromKVO {
+  if (!self.textInput || !self.isRegisteredForKVO) {
+    return;
+  }
+  @try {
+    [self.textInput.leadingUnderlineLabel removeObserver:self
+                                              forKeyPath:MDCTextInputControllerKVOKeyFont];
+    [self.textInput.placeholderLabel removeObserver:self
+                                         forKeyPath:MDCTextInputControllerKVOKeyFont];
+    [self.textInput.trailingUnderlineLabel removeObserver:self
+                                               forKeyPath:MDCTextInputControllerKVOKeyFont];
+  } @catch (NSException *exception) {
+  }
+  _isRegisteredForKVO = NO;
+}
+
+#pragma mark - Character Max Implementation
+
+- (NSUInteger)characterCount {
+  return [self.characterCounter characterCountForTextInput:self.textInput];
+}
+
+- (id<MDCTextInputCharacterCounter>)characterCounter {
+  if (!_characterCounter) {
+    _characterCounter = self.internalCharacterCounter;
+  }
+  return _characterCounter;
+}
+
+- (void)setCharacterCounter:(id<MDCTextInputCharacterCounter>)characterCounter {
+  if (_characterCounter != characterCounter) {
+    _characterCounter = characterCounter;
+    [self updateLayout];
+  }
+}
+
+- (void)setCharacterCountMax:(NSUInteger)characterCountMax {
+  if (_characterCountMax != characterCountMax) {
+    _characterCountMax = characterCountMax;
+    [self updateLayout];
+  }
+}
+
+#pragma mark - Leading Label Customization
+
+- (void)updateLeadingUnderlineLabel {
+  if (self.presentationStyle == MDCTextInputPresentationStyleFullWidth) {
+    self.textInput.leadingUnderlineLabel.text = nil;
+    return;
+  }
+
+  if (!self.customLeadingFont) {
+    self.textInput.leadingUnderlineLabel.font = [[self class] underlineLabelsFont];
+  }
+
+  self.textInput.leadingUnderlineLabel.textColor =
+      (self.isDisplayingErrorText || self.isDisplayingCharacterCountError)
+          ? self.errorColor
+          : MDCTextInputInlinePlaceholderTextColor();
+}
+
+#pragma mark - Placeholder Customization
+
+- (void)updatePlaceholder {
+  if (!self.customPlaceholderFont) {
+    self.textInput.placeholderLabel.font = [[self class] placeholderFont];
+  }
+}
+
+- (BOOL)isPlaceholderUp {
+  return self.placeholderAnimationConstraints.count > 0 &&
+         !CGAffineTransformEqualToTransform(self.textInput.placeholderLabel.transform,
+                                            CGAffineTransformIdentity);
+}
+
+#pragma mark - Placeholder Animation
+
+- (void)movePlaceholderToUp:(BOOL)isToUp {
+  if (self.isPlaceholderUp && isToUp) {
+    return;
+  }
+
+  if (!self.isPlaceholderUp && !isToUp) {
+    return;
+  }
```

Comment:
For the two if statements above, I typically see this as:
if (self.isPlaceholderUp == isToUp) {
  return;
}